### PR TITLE
[Part 2] Adding s390x cross-compilation support for gcr.io images in this repo

### LINKED ISSUE #34328: Regarding required changes for s390x architecture
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): Yes

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):s390x

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

  I am working on Kubernetes for s390x architecture to enable docker images for s390x should be pushed to gcr.io repository. For this
1. I was trying to build the master branch but it failed for s390x due to there was an issue in the file "server_statvfs_linux.go". The thing is this issue is resolved in the sftp package but the package is not updated in the Kubernetes repository.  Can you give me an idea when this package will be updated in the Kubernetes code(vendor/github.com/pkg/sftp) ?
2. Also for the same task when I ran build/release.sh command, I came across docker image gcr.io/google_containers/kube-cross which is using golang:1.6.3(https://github.com/kubernetes/kubernetes/blob/master/build/build-image/cross/Dockerfile) as base image and it does not support 
   s390x architecture. Go 1.7.1 is released and it has support for s390x architecture. Is there any plan to update the mentioned docker file for golang 1.7.1 so it will support multiple architectures supported by Go?? And if there is no plan can I add my Docker file specific to s390x so that gcr.io/google_containers/kube-cross docker image will have support for s390x.

@cwsolee  @aborkar-ibm

### PR DESCRIPTION
**Pull request by gajju26**
**What this PR does / why we need it**: This PR enables s390x support to kube-dns , pause, addon-manager, etcd, hyperkube, kube-discovery etc. This PR also includes the changes due to which it can be cross compiled on x86 host architecture.

**Which issue this PR fixes #34328

**Special notes for your reviewer**:  In existing file "build-tools/build-image/cross/Dockerfile" the repository mentioned for installing cross build tool chains for supporting architecture does not have a tool chain for s390x hence in my PR I am changing the repository so that it will be cross compiled for s390x.

**Release note**:

```
Allows cross compilation of Kubernetes on x86 host for s390x also enables s390x support to kube-dns , pause, addon-manager, etcd, hyperkube, kube-discovery etc  
```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36050)

### CODE REVIEW

**Comment by dims** [comment]
@k8s-bot ok to test

**Comment by ixdy** [comment]
@k8s-bot build this

**Comment by ixdy** [comment]
cc @luxas

File: build-tools/build-image/cross/Dockerfile
Comment (ixdy) [inline build-tools/build-image/cross/Dockerfile]: why was this line removed?

File: build-tools/build-image/cross/Dockerfile
Comment (ixdy) [inline build-tools/build-image/cross/Dockerfile]: we probably need to split this PR into two parts, since we need a new cross-build (and other) images before the cross-build will work in Jenkins.

File: vendor/github.com/opencontainers/runc/libcontainer/system/syscall_linux_64.go
Comment (ixdy) [inline vendor/github.com/opencontainers/runc/libcontainer/system/syscall_linux_64.go]: did this change upstream? we shouldn't be changing things in vendor/ without a corresponding Godeps change.

**Review by ixdy — COMMENTED** [review]
This roughly seems fine, though as I've noted we probably need to split this into at least 2 PRs so that we can push necessary images first.

You'll also want to update `cluster/get-kube-binaries.sh` to include support for s390x.

And please clean up the merge commits on this PR, too.

File: build-tools/build-image/cross/Dockerfile
Comment (luxas) [inline build-tools/build-image/cross/Dockerfile]: I think this is a pretty big question of trust. Do we trust this repository?
Is it possible to separate the others to use emdebian and only make s390x use something other?

File: build-tools/build-image/cross/Dockerfile
Comment (luxas) [inline build-tools/build-image/cross/Dockerfile]: that line should be kept

File: build-tools/build-image/cross/Dockerfile
Comment (luxas) [inline build-tools/build-image/cross/Dockerfile]: nit: unnecessary change

File: cluster/addons/addon-manager/Makefile
Comment (luxas) [inline cluster/addons/addon-manager/Makefile]: why v1.3.0-beta.2?

File: vendor/github.com/opencontainers/runc/libcontainer/system/syscall_linux_64.go
Comment (luxas) [inline vendor/github.com/opencontainers/runc/libcontainer/system/syscall_linux_64.go]: also that change seems unnecessary...

**Review by luxas — COMMENTED** [review]
Very cool!
My goal with the multiarch refactoring I made earlier was to make this possible easily, and here we go!

But I'm on the same line as @ixdy, I'd like us to start building binaries first and making sure they work before pushing a lot of images.

So please split out the changes you've made to two PRs: one that adds support to `kube-cross` to cross-compile the binaries etc. and one that updates all the images to include a s390x variant.

Again, thanks for this PR

**Comment by luxas** [comment]
You should also sign the linux foundation CLA please

**Comment by gajju26** [comment]
@luxas @ixdy  Thanks for reviewing this PR. I am working on the review comments of this PR. Can I get some information about CLA of linuxfoundation.

**Comment by ixdy** [comment]
@gajju26 if you follow the [Details](https://identity.linuxfoundation.org/projects/cncf) link it should tell you about the CLA. You can also read the agreements [here](https://github.com/cncf/cla).

**Comment by ixdy** [comment]
also FYI - the 1.5 merge window is closing, and this PR seems unlikely to make the cutoff. I imagine it will make it in soon after the code freeze thaws, however.

File: build-tools/build-image/cross/Dockerfile
Comment (gajju26) [inline build-tools/build-image/cross/Dockerfile]: @luxas @ixdy  I have been scratching my head to resolve this problem. Actually with the existing (jessie main) repository, there is no crossbuild-essential-s390x available. For cross compilation we need s390x-linux-gnu-gcc binary . I tried your suggested approach keeping the "jessie main" for existing supported architectures and introducing new repo for only for s390x but this solution is not working. Keeping two repositories cause dependencies and version issues error for the packages (like libc and glibc) which are used by s390x and existing architecture. I think there can be 2 ways to address this problem.
1. We can upgrade the repository to xenial as I did in the PR. I think this approach is quite feasible because having the latest stable repository  will not break anything also it will provide higher versions of packages and down the line we will have to upgrade it to latest repositories. 
2. For s390x only I can include the steps for cross compiling gcc from the ource so that it would create s390x-linux-gnu-gcc binary and which will be used for compiling Kubernetes on s390x. Once we upgrade or enable the proper repository we would remove this this step. 

I will need your inputs to decide a better solution.

Comment (ixdy) [inline build-tools/build-image/cross/Dockerfile]: yeah, I'm not sure there's a great way to handle this without using ubuntu xenial. It seems like the s390x crossbuild just isn't available in the regular debian repos.

minor suggestion: we might want to use a generic ubuntu mirror rather than pinning to the Czech one.

File: build-tools/build-image/cross/Dockerfile
Comment (ixdy) [inline build-tools/build-image/cross/Dockerfile]: maybe update this comment, since we aren't using the debian crosstools?

**Comment by gajju26** [comment]
Hi @ixdy  @luxas 

Making another PR for "build-tools/build-image/cross/Dockerfile" to include s390x changes is in progress. I will submit  that PR soon.

**Comment by gajju26** [comment]
PR for "build-tools/build-image/cross/Dockerfile"
https://github.com/kubernetes/kubernetes/pull/37092

**Comment by gajju26** [comment]
Hi luxas,
I have signed "the linux foundation CLA" with same github user id i.e. gajju26.

**Comment by gajju26** [comment]
Hi @ixdy  @luxas ,

Since #37092  is closed can you have a look on this PR? I will fix if there is any issue.

**Comment by ixdy** [comment]
@k8s-bot build this
@k8s-bot test this

File: test/images/serve_hostname/Makefile
Comment (ixdy) [inline test/images/serve_hostname/Makefile]: nit: move "and" after ppc64le

**Review by ixdy — APPROVED** [review]
LGTM. @luxas if you think this is good, feel free to tag.

**Comment by ixdy** [comment]
@gajju26 can you update the PR description and fill in the release note block, please?

File: hack/lib/golang.sh
Comment (ixdy) [inline hack/lib/golang.sh]: oh, this `fi` needs to be removed.

might be worth turning this massive `if/elif/elif/elif/...` block into a `case` block.

**Review by ixdy — CHANGES_REQUESTED** [review]

File: hack/lib/golang.sh
Comment (gajju26) [inline hack/lib/golang.sh]: Replaced by case block.

File: cluster/addons/addon-manager/Makefile
Comment (luxas) [inline cluster/addons/addon-manager/Makefile]: Why not `s390x/debian` like the other arches?
If you want to refactor this then to using other base images, it's fine, but not in this PR

File: hack/lib/golang.sh
Comment (luxas) [inline hack/lib/golang.sh]: nit: `Unsupported architecture to cross-compile from linux/amd64`

**Review by luxas — COMMENTED** [review]
LGTM except for some nits, please fix them

After that, it's up to @ixdy to actually push the images once this merges.

By giving LGTM, I assume you've built all images locally and tested them so they actually work. You've done this, right?

(Also, I'm specifically interested in if Qemu supports emulating `s390x` properly)

File: hack/lib/golang.sh
Comment (ixdy) [inline hack/lib/golang.sh]: there should probably be a case for "linux/amd64" which does nothing. right now this is failing in CI when trying to target `linux/amd64`.

(alternately, update the `if` statement on L234 to handle this case.)

File: cluster/addons/addon-manager/Makefile
Comment (ixdy) [inline cluster/addons/addon-manager/Makefile]: how is `QEMUARCH` used? the other arches here don't use it.

Comment (gajju26) [inline cluster/addons/addon-manager/Makefile]: Actually I had to mention "s390x/debian" only but due to some wrong commit merges it got changed.

File: hack/lib/golang.sh
Comment (gajju26) [inline hack/lib/golang.sh]: @ixdy I have added a case statement for "linux/amd64" which does nothing. In my opinion after adding this case statement, no need to modify the if condition on L234 right? Bit confused.

**Review by ixdy — APPROVED** [review]

File: hack/lib/golang.sh
Comment (ixdy) [inline hack/lib/golang.sh]: yeah, it was an either-or. this is fine as-is.

**Review by luxas — APPROVED** [review]
LGTM to me as well :+1:

**Comment by ixdy** [comment]
FYI I'd forgotten to push images for s390x, so looking into that now.
Pushed:
* `gcr.io/google_containers/debian-iptables-s390x:v5`
* `gcr.io/google_containers/pause-s390x:3.0`
* `gcr.io/google-containers/kube-addon-manager-s390x:v6.1`

Issues:
* kubedns: @bowei can you handle this, since it moved to a different repo? also it looks like dnsmasq (https://github.com/kubernetes/dns/blob/master/images/dnsmasq/Makefile) has no s390x support yet.
* `cluster/images/etcd` uses go1.6.3, so we can't currently build `gcr.io/google_containers/etcd-s390x`
* `google_containers/hyperkube-s390x`: I suspect this will automatically be released with 1.6. nothing there yet, though.
* `cluster/images/kube-discovery` has weird versioning - its tag is just "1.0", but it depends on whatever's been built in the tree, so I'm not sure what commit to push for s390x. @dgoodwin can maybe help here.
* `gcr.io/google_containers/serve_hostname` - it looks like nobody has pushed any non-amd64 images for this one yet.

**Comment by luxas** [comment]
I'm pretty optimistic I have https://github.com/kubernetes/kubernetes/pull/38926 working now, building atm

**Pull request merged**
